[v3 backport] Prevent push cmd failure in 3.18 by handling version tag resolution in ORAS memory store

benoittgt opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/helm/helm/blob/main/CONTRIBUTING.md
2. If this PR closes another issue, add 'closes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->

**What this PR does / why we need it**:

Fix: https://github.com/helm/helm/issues/30881

Backport of https://github.com/helm/helm/pull/30894

**Special notes for your reviewer**:

**If applicable**:
- [ ] this PR contains user facing changes (the `docs needed` label should be applied if so)
- [ ] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

benoittgt committed:
Prevent failure when resolving version tags in oras memory store

- The newReference() function transforms version tags by replacing + with _ for OCI compatibility
- But the code was using the original ref (with +) for TagBytes()
- Then it tries to find the tagged reference using parsedRef.String() (with _)
- This mismatch causes the Resolve method to fail with "not found"
- By using parsedRef.String() consistently in both places, the references will match and the lookup will succeed.

Close: https://github.com/helm/helm/issues/30881
Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

gjenkins8 commented:
@benoittgt -- could you please kindly target the `main` branch with this PR (so the fix also gets applied to main). Then we will backport to `dev-v3` branch.

Also, if you have time to add a test. That would be much appreciated.

benoittgt committed:
Prevent failure when resolving version tags in oras memory store

- The newReference() function transforms version tags by replacing + with _ for OCI compatibility
- But the code was using the original ref (with +) for TagBytes()
- Then it tries to find the tagged reference using parsedRef.String() (with _)
- This mismatch causes the Resolve method to fail with "not found"
- By using parsedRef.String() consistently in both places, the references will match and the lookup will succeed.

I extracted the TagBytes function to improve testability.
Push() includes several external calls that are hard to mock,
so isolating this logic makes testing more manageable.

Close: #30881
Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt commented:
@gjenkins8 I did https://github.com/helm/helm/pull/30894 and I will make the packport when approved 😊. Sorry I often forgot.

benoittgt committed:
Revert "Prevent failure when resolving version tags in oras memory store"

This reverts commit 10427fc0e4637e677f293fcd2a4aadeca0dc2c83.

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

mattfarina reviewed (commented):
@benoittgt Thanks for the backport. It works but there is a conflict. Can you please resolve it.

benoittgt commented:
We will probably prefer https://github.com/helm/helm/pull/30906. So closing.
